connector/src/main/java/org/geysermc/connector/network/translators/java/world/JavaBlockValueTranslator.java

```diff
@@ -62,23 +63,42 @@ public void translate(GeyserSession session, ServerBlockValuePacket packet) {
         } else if (packet.getValue() instanceof NoteBlockValue) {
             NoteblockBlockEntityTranslator.translate(session, packet.getPosition());
         } else if (packet.getValue() instanceof PistonValue) {
-            if (session.getConnector().getPlatformType() == PlatformType.SPIGOT) {
-                // Handled in the GeyserPistonEvents class
-                return;
-            }
-
             PistonValueType action = (PistonValueType) packet.getType();
-            PistonValue direction = (PistonValue) packet.getValue();
+            Direction direction = Direction.fromPistonValue((PistonValue) packet.getValue());
             Vector3i position = Vector3i.from(packet.getPosition().getX(), packet.getPosition().getY(), packet.getPosition().getZ());
-
             PistonCache pistonCache = session.getPistonCache();
-            PistonBlockEntity blockEntity = pistonCache.getPistons().computeIfAbsent(position, pos -> {
-                int blockId = session.getConnector().getWorldManager().getBlockAt(session, position);
-                boolean sticky = BlockStateValues.isStickyPiston(blockId);
-                boolean extended = action != PistonValueType.PUSHING;
-                return new PistonBlockEntity(session, pos, Direction.fromPistonValue(direction), sticky, extended);
-            });
-            blockEntity.setAction(action);
+
+            if (session.getConnector().getPlatformType() == PlatformType.SPIGOT) {
+                // Mostly handled in the GeyserPistonEvents class
+                // Retracting sticky pistons is an exception, since the event is not called on Spigot from 1.13.2 - 1.17.1
+                // See https://github.com/PaperMC/Paper/blob/master/patches/server/0304-Fire-BlockPistonRetractEvent-for-all-empty-pistons.patch
```

Comment:
Set this to a static commit instead of just master; patch numbers tend to shift around quite a bit.